tabular/src/autogluon/tabular/predictor/predictor.py

```diff
@@ -3299,6 +3319,93 @@ def explain_classification_errors(self, data, model = None, print_rules: bool =
         cls, columns = imodels.explain_classification_errors(data, predictions, labels, print_rules=print_rules)
         return cls
 
+    # TODO: Add .delete() method to easily clean-up clones?
+    #  Would need to be careful that user doesn't delete important things accidentally.
+    # TODO: Add .save_zip() and load_zip() methods to pack and unpack artifacts into a single file to simplify deployment code?
+    def clone(self,
+              path: str,
+              *,
+              return_clone: bool = False,
+              dirs_exist_ok: bool = False):
+        """
+        Clone the predictor and all of its artifacts to a new location on local disk.
+        This is ideal for use-cases where saving a snapshot of the predictor is desired before performing
+        more advanced operations (such as fit_extra and refit_full).
+
+        Parameters
+        ----------
+        path : str
+            Directory path the cloned predictor will be saved to.
+        return_clone : bool, default = False
+            If True, returns the loaded cloned TabularPredictor object.
+            If False, returns the local path to the cloned TabularPredictor object.
+        dirs_exist_ok : bool, default = False
+            If True, will clone the predictor even if the path directory already exists, potentially overwriting unrelated files.
+            If False, will raise an exception if the path directory already exists and avoid performing the copy.
+
+        Returns
+        -------
+        If return_clone == True, returns the loaded cloned TabularPredictor object.
+        If return_clone == False, returns the local path to the cloned TabularPredictor object.
```

Comment:
Having the return type change based on this parameter seems like a code smell to me. But I don't have a good solution/suggestion here, so I'm just calling it out to see if anyone else had thoughts.

Reply:
Could you explain why it is a code smell? What situation would this lead to suboptimal functionality / usage?

Reply:
If we were using a statically typed language, we would need to return either the cloned object or a path. In some languages you could do something fancy with the return type like `-> Union[TabularPredictor, PathStr]`, but often this kind of return type handling reveals a bad interface. Obviously since we're using Python, we don't have the limitations a statically typed language, but I think that perspective is still valid. My concern is if a user has logic that expects a TabularPredictor, then they decide to change this parameter and suddenly they are working with string, they would be surprised even if the documentation spells it out clearly.

An alternative solution is to have clone always return the path string, and users have to call `load()` separately if they want to bring the cloned TabularPredictor into memory. That seems like a good separation of concerns despite being a bit less convenient.

I just wanted to surface this as an outside observer. If you are happy with this interface, I don't oppose it.

Reply:
I understand where you are coming from, and it is definitely a reasonable perspective. I think that so long as the separation of concerns is only broken if they opt-in to use the convenience argument, then it is fine. I foresee that if this argument didn't exist, the two lines of code would be very common, so I like to simplify this down to 1 line of code when I anticipate it will be the primary use-case. You could argue our `predict_proba` having `as_pandas` argument is the same situation. There is no objectively right answer here, but I don't see the harm.